### DISCUSSION TITLE
A proper pointcloud renderer like the one in ARCore?

### DISCUSSION THREAD

**Discussion by sarimmehdi** [sceneview/sceneform-android#278] [Ideas]
Hello. I understand that there are other more important features that take priority but I was hoping for a PointCloud renderer to be integrated into SceneForm. 

Currently, I manually project the point cloud onto the image using the camera projection matrix for each new frame. Since there are many points for each frame, I perform the task of computing the projections of all the points on a separate non-UI thread and then draw the 2D points on a custom view that I have overlaid on top of ArSceneView.

I get the feeling that this process could probably be done faster? In ARCore they use OpenGL to do this, as described [here](https://chromium.googlesource.com/external/github.com/google-ar/arcore-android-sdk/+/refs/tags/v1.2.0/samples/cloud_anchor_java/app/src/main/java/com/google/ar/core/examples/java/common/rendering/PointCloudRenderer.java). 

Right now, I just do the whole thing in Java. I create arrays out of the camera intrinsics matrix and projection matrix. Then, I create a separate array that represents the result obtained by multiplying the two to get the projection matrix and then multiplying that with a 4 by n matrix of points to get their projection. Is there a way this whole process could be integrated into Sceneform? I know there is a Matrix class but I was actually thinking of going the C++ route to make the whole thing super fast (probably with the help of a third-party linear algebra library).

**Comment by grassydragon** [comment]
Hi!
I have implemented the point cloud rendering using Filament and dynamically adding and removing the points from the scene. I render the points using unlit transparent spheres that look like points on the screen. I haven't encountered any performance issues with this implementation.
![Point Cloud](https://user-images.githubusercontent.com/35110066/148892120-048ae431-9915-4daf-977e-91512f072649.png)

**Reply by sarimmehdi** [reply]
Can you please share your code? I do not know how to use filament. Maybe you can also commit this to the repo so that others can use it as well? Thanks

**Reply by grassydragon** [reply]
I think, yes, it is not very hard to include it in the repository.

**Reply by grassydragon** [reply]
`PointRenderer` class:
``` Kotlin
import android.content.Context
import android.opengl.Matrix
import com.google.android.filament.*
import com.google.ar.core.Frame
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

class PointRenderer(
    context: Context,
    private val engine: Engine,
    private val scene: Scene
) {

    companion object {

        private const val RADIUS = 0.005f

        private const val VERTICAL_SEGMENTS_COUNT = 10

        private const val HORIZONTAL_SEGMENTS_COUNT = 20

    }

    private val renderables: ArrayList<Int> = ArrayList()

    private val material: Material

    private val vertexBuffer: VertexBuffer

    private val indexBuffer: IndexBuffer

    private val indexCount: Int

    private var lastUpdateTime = 0L

    private var pointsCount = 0

    var enabled: Boolean = false
        set(value) {
            if (value != field) {
                field = value

                if (value) {
                    for (i in 0 until pointsCount) scene.addEntity(renderables[i])
                } else {
                    for (i in 0 until pointsCount) scene.removeEntity(renderables[i])
                }
            }
        }

    init {
        material = loadMaterial(context, engine, "point_renderer.filamat")

        val m = VERTICAL_SEGMENTS_COUNT
        val n = HORIZONTAL_SEGMENTS_COUNT

        val floatSize = 4
        val shortSize = 2

        val vertexSize = 3 * floatSize

        val vertexCount = (m - 1) * n + 2

        indexCount = (m - 1) * n * 2 * 3

        val vertexData = ByteBuffer.allocate(vertexCount * vertexSize).order(ByteOrder.nativeOrder())

        vertexData.putVertex(0f, RADIUS, 0f)

        for (i in 1 until m) {
            val a = PI.toFloat() * i / m

            val h = RADIUS * cos(a)
            val l = RADIUS * sin(a)

            for (j in 0 until n) {
                val b = PI.toFloat() * 2 * j / n

                vertexData.putVertex(l * sin(b), h, l * cos(b))
            }
        }

        vertexData.putVertex(0f, -RADIUS, 0f)

        vertexData.flip()

        vertexBuffer = VertexBuffer.Builder()
            .bufferCount(1)
            .vertexCount(vertexCount)
            .attribute(VertexBuffer.VertexAttribute.POSITION, 0, VertexBuffer.AttributeType.FLOAT3, 0, vertexSize)
            .build(engine)

        vertexBuffer.setBufferAt(engine, 0, vertexData)

        val indexData = ByteBuffer.allocate(indexCount * shortSize).order(ByteOrder.nativeOrder())

        for (i in 0 until n) {
            indexData.putTriangle(0, i + 1, (i + 1) % n + 1)
        }

        for (i in 0 until m - 2) {
            for (j in 0 until n) {
                val left = i * n + j + 1
                val right = i * n + (j + 1) % n + 1

                indexData.putTriangle(left, left + n, right)
                indexData.putTriangle(right, left + n, right + n)
            }
        }

        for (i in 0 until n) {
            indexData.putTriangle((m - 1) * n + 1, (i + 1) % n + (m - 2) * n + 1, i + (m - 2) * n + 1)
        }

        indexData.flip()

        indexBuffer = IndexBuffer.Builder()
            .indexCount(indexCount)
            .bufferType(IndexBuffer.Builder.IndexType.USHORT)
            .build(engine)

        indexBuffer.setBuffer(engine, indexData)
    }

    fun destroy() {
        engine.apply {
            renderables.forEach { destroyEntity(it) }
            destroyMaterial(material)
            destroyVertexBuffer(vertexBuffer)
            destroyIndexBuffer(indexBuffer)
        }
    }

    fun update(frame: Frame) {
        if (!enabled) return

        frame.acquirePointCloud().use { pointCloud ->
            if (pointCloud.timestamp != lastUpdateTime) {
                lastUpdateTime = pointCloud.timestamp

                val newPointCount = pointCloud.points.limit() / 4

                for (i in renderables.size until newPointCount) createRenderable()

                if (newPointCount < pointsCount) {
                    for (i in newPointCount until pointsCount) scene.removeEntity(renderables[i])
                }
                else {
                    for (i in pointsCount until newPointCount) scene.addEntity(renderables[i])
                }

                pointsCount = newPointCount

                for (i in 0 until pointsCount) {
                    val x = pointCloud.points[i * 4]
                    val y = pointCloud.points[i * 4 + 1]
                    val z = pointCloud.points[i * 4 + 2]

                    val matrix = FloatArray(16)

                    Matrix.setIdentityM(matrix, 0)
                    Matrix.translateM(matrix, 0, x, y, z)

                    val transformManager = engine.transformManager
                    transformManager.setTransform(transformManager.getInstance(renderables[i]), matrix)
                }
            }
        }
    }

    private fun createRenderable() {
        @Entity
        val renderable = EntityManager.get().create()

        RenderableManager.Builder(1)
            .boundingBox(Box(0f, 0f, 0f, RADIUS, RADIUS, RADIUS))
            .geometry(0, RenderableManager.PrimitiveType.TRIANGLES, vertexBuffer, indexBuffer, 0, indexCount)
            .material(0, material.defaultInstance)
            .build(engine, renderable)

        renderables.add(renderable)
    }

}
```
`loadMaterial` method:
``` Kotlin
import android.content.Context
import com.google.android.filament.Engine
import com.google.android.filament.Material
import java.nio.ByteBuffer
import java.nio.channels.Channels

fun loadMaterial(context: Context, engine: Engine, name: String): Material {
    val dst: ByteBuffer

    context.assets.openFd("materials/$name").use { fd ->
        val input = fd.createInputStream()
        dst = ByteBuffer.allocate(fd.length.toInt())

        val src = Channels.newChannel(input)
        src.read(dst)
        src.close()

        dst.rewind()
    }

    return Material.Builder().payload(dst, dst.remaining()).build(engine)
}
```
`ByteBuffer` extensions:
``` Kotlin
import java.nio.ByteBuffer

fun ByteBuffer.putVertex(x: Float, y: Float, z: Float): ByteBuffer {
    putFloat(x)
    putFloat(y)
    putFloat(z)

    return this
}

fun ByteBuffer.putTriangle(first: Int, second: Int, third: Int): ByteBuffer {
    putShort(first)
    putShort(second)
    putShort(third)

    return this
}
```

**Reply by grassydragon** [reply]
@ThomasGorisse @RGregat
Should we add point cloud rendering to SceneView?

**Reply by RGregat** [reply]
To be honest, I thought about it already. Same for some more primitives/simple models. They are handy for sample projects where you don't want to bother with 3D-Models.
Btw I have a version with pyramids to reduce the vertex/triangle/normal count

**Reply by ThomasGorisse** [reply]
For sure if you have a usage type sample because I don't see any other than for testing purposes.
Just like it's said in the ARCore doc:
[public PointCloud acquirePointCloud()](https://developers.google.com/ar/reference/java/com/google/ar/core/Frame#acquirePointCloud-)

> Note: This information is for visualization and debugging purposes only. Its characteristics and format are subject to change in subsequent versions of the API.

This is not a "we shouldn't" but a "why we should?"

**Reply by sarimmehdi** [reply]
I imagine many people like me would be using this for research purposes and it would be handy for them to be able to plot a point cloud right out of the box without spending some time implementing it themselves.

**Reply by grassydragon** [reply]
> For sure if you have a usage type sample because I don't see any other than for testing purposes.

I have only seen this in Measure by Google that unfortunately is no longer available. The point cloud is shown for a short time after the session comes to the tracking state. This creates a visual effect of the phone recognizing the real world environment.

**Reply by ThomasGorisse** [reply]
Ok. Let's put it in something like the PlaneVisualizer which can be enabled at the ArSceneView level.

`arSceneView.pointCloudVisible: Boolean = false`?

**Reply by sarimmehdi** [reply]
Hello. Just wanted to check. Has this been integrated into the current version of Sceneform?

**Reply by grassydragon** [reply]
Hi!
Sorry I didn't have time to do that and probably we will integrate it into SceneView since all new features will be appearing there.
